libs/chain-signatures/contract/src/lib.rs

```diff
@@ -62,70 +70,82 @@ impl Default for VersionedMpcContract {
     }
 }
 
-pub trait SignatureStateHandler {
-    fn add_request(&mut self, request: &SignatureRequest, data_id: CryptoHash);
-    fn get_pending_request(&mut self, request: &SignatureRequest) -> Option<YieldIndex>;
-    fn remove_request(&mut self, request: SignatureRequest) -> Result<(), Error>;
+#[derive(BorshDeserialize, BorshSerialize, Debug)]
+pub struct MpcContractV2 {
+    protocol_state: ProtocolContractState,
+    pending_requests: LookupMap<SignatureRequest, YieldIndex>,
+    request_by_timestamp: Vector<(u64, SignatureRequest)>,
+    proposed_updates: ProposedUpdates,
+    request_timeout: u64,
 }
 
-pub fn handle_signature_response<T: SignatureStateHandler>(
-    mpc_contract: &mut T,
-    request: &SignatureRequest,
-    response: &SignatureResponse,
-) -> Result<(), Error> {
-    if let Some(YieldIndex { data_id }) = mpc_contract.get_pending_request(request) {
-        env::promise_yield_resume(&data_id, &serde_json::to_vec(&response).unwrap());
-        Ok(())
-    } else {
-        Err(InvalidParameters::RequestNotFound.into())
+impl MpcContractV2 {
+    fn remove_timed_out_requests(&mut self) {
+        let deadline = env::block_timestamp_ms() - self.request_timeout;
```

Comment:
The yielded promise created by the sign call [times out after 200 blocks](https://github.com/near/nearcore/blob/b1bb76ce2e4dd53159c04830ef69e86d4663281e/core/parameters/res/runtime_configs/parameters.snap#L197). You could handle the timeouts accordingly using `env::block_height` instead of `env::block_timestamp_ms`.